aleph/model/role.py

```diff
@@ -118,8 +118,11 @@ def clear_roles(self):
 
     def add_role(self, role):
         """Adds an existing role as a membership of a group."""
-        self.roles.append(role)
-        db.session.add(role)
+        if role not in self.roles:
```

Comment:
I wonder if catching https://docs.sqlalchemy.org/en/20/core/exceptions.html#sqlalchemy.exc.IntegrityError here and reacting on it (log.warn) would be more efficient than what I assume is a SELECT followed by an INSERT?

Reply:
This operates on the ORM object only, i.e., no requests are made at this time. Only after flushing the SQLAlchemy session and comitting the transaction statements to update the role memberships would be executed (and any errors would be raised only then).

Reply:
The primary goal of this log message was to confirm that there is no logic flaw that can insert duplicate role memberships within a single request. I’m pretty sure there is no such flaw (instead the issue is a race condition with parallel requests) and thus we should never see this log message in practice. But I’m not 100% sure.